[core]/es_extended/shared/functions.lua

```
@@ -170,4 +170,52 @@ function ESX.IsFunctionReference(val)
     local typeVal = type(val)
 
     return typeVal == "function" or (typeVal == "table" and type(getmetatable(val)?.__call) == "function")
+end
+
+---@param conditionFunc function A function that returns a boolean indicating whether the condition is met.
+---@param errorMessage? string The error message to print if the condition is not met within the timeout period.
+---@param timeout? number The maximum time (in milliseconds) to wait for the condition to be met.
+---@return boolean, number: Returns success status and the time taken
+ESX.Await = function(conditionFunc, errorMessage, timeout)
+    timeout = timeout or 1000
+    
+    if timeout < 0 then
+        error('Timeout should be a positive number.')
+    end
+
+    local isFunctionReference = ESX.IsFunctionReference(conditionFunc)
+
+    if not isFunctionReference then
+        error('Condition Function should be a function reference.')
+    end
+
+    local startTime = GetGameTimer()
+    local prefix = ('[%s] -> '):format(GetInvokingResource())
+
+    while GetGameTimer() - startTime < timeout do
+        local success, result = pcall(conditionFunc)
+
+        if not success then
+            local conditionErrorMessage = ('%s Error while calling conditionFunc! Result: %s'):format(prefix, result)
+            print(conditionErrorMessage)
+            local elapsedTime = GetGameTimer() - startTime
+            return false, elapsedTime
+        end
+
+        if result then
+            local elapsedTime = GetGameTimer() - startTime
+            return true, elapsedTime
+        end
+
+        Wait(0)
+    end
+
+    if errorMessage then
+        ESX.AssertType(errorMessage, 'string', 'errorMessage should be a string.')
+
+        local formattedErrorMessage = ('%s %s'):format(prefix, errorMessage)
+        print(formattedErrorMessage)
```

We should throw an actual error here.